[iOS SDK]Universal Bridge : Communication between the Host App and MiniApps

GondiTilak opened the pull request:
# Description
Universal Bridge feature implementation in iOS SDK:
Any MiniApp can send message to host app using sendJsonToHostApp action, this will be received in SDK delegate method  `sendJsonToHostApp(info: , completionHandler:)`.
Any MiniApp can receive message from the host app using the method in SDK func `sendJsonToMiniApp(string jsonString:)`

## Links
[MINI-5752](https://jira.rakuten-it.com/jira/browse/MINI-5752)
[MINI-5754](https://jira.rakuten-it.com/jira/browse/MINI-5754)
[Feature Demo Video](https://rak.box.com/s/tic8g6stv1bgvw49fo92cnbea45bl5ha)

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `fastlane ci` without errors

khairul-alam-licon commented on Sources/Classes/core/MiniApp.swift:
```diff
@@ -204,6 +204,17 @@ public class MiniApp: NSObject {
+    ///   - miniAppId: Mini AppId String value
+    ///   - miniAppVersion: Mini VersionId String value
+    ///   - jsonString: string to be snt to MiniApp
+    public func sendJsonStringToMiniApp(withId miniAppId: String,
```
[In Android](https://github.com/rakutentech/android-miniapp/blob/master/miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppView.kt#L73), we have written `sendJsonToMiniApp`. Probably, we can align the same name for the public interfaces.

rleojoseph replied:
@GondiTilak , please change the method name from 'sendJsonStringToMiniApp' to 'sendJsonToMiniApp' as per the tech spec. Also, I don't think this method is required in `MiniApp.swift` class as these interfaces is have shared config.

rleojoseph replied:
Please create interfaces in MiniAppView class like in Android

rleojoseph replied:
Could you please remove this method?

GondiTilak replied:
This is now removed in the latest commit.

rleojoseph commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppBridgeView.swift:
```diff
@@ -0,0 +1,41 @@
+import SwiftUI
+import MiniApp
+
+struct MiniAppBridgeView: View {
```
MiniAppBridgeView is too generic that it could mean the entire bridge used to communicate MiniApp. May be "MessageBridgeView" or "UniversalBridgeView"?

GondiTilak replied:
Sure got the point, I have updated to use the name `UniversalBridgeView` also update the relevant class file name.

rleojoseph commented on Example/Controllers/SwiftUI/Features/Terms/MiniAppWithTermsView.swift:
```diff
@@ -35,12 +36,21 @@ struct MiniAppWithTermsView: View {
-            }))
+            if let message = viewModel.showJsonString {
+                return Alert(
+                    title: Text("Universal Bridge"),
```
Is this the message received from MiniApp ?

GondiTilak replied:
Yes this is the message received from miniapp for interface1 implementation.

rleojoseph commented on Example/Controllers/SwiftUI/Helper/MiniAppViewMessageDelegator.swift:
```diff
@@ -254,3 +263,9 @@ extension MiniAppViewMessageDelegator: CLLocationManagerDelegate {
+
+extension MiniAppViewMessageDelegator {
+    func sendJsonStringToMiniApp(jsonString: String) {
+        MiniApp.shared().sendJsonStringToMiniApp(withId: miniAppId, miniAppVersion: miniAppVersion ?? "", jsonString: jsonString)
```
We shouldn't support this. Host app can't use our old singleton patterns to send message to Mini App

GondiTilak replied:
This is now removed in the latest commits.

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+extension MiniAppScriptMessageHandler {
+    func sendJsonToHostApp(requestParam: RequestParameters?, callbackId: String) {
+        guard let requestParamValue = requestParam?.jsonInfo else {
+            executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: MiniAppJavaScriptError.invalidPermissionType.rawValue)
```
We don't need any permissions here. So throwing a permission related error here is not required.
Please use 'prepareMAJavascriptError(MiniAppJavaScriptError.unexpectedMessageFormat)' or any other error

GondiTilak replied:
For this I have now created new error enum object `UniversalBridgeError` to cater for universal bridge related errors. With this I can now use the `prepareMAJavascriptError(error)`

rleojoseph replied:
@GondiTilak . No, we don't need any custom permissions to use this interface. Throwing an error related to invalid permission is misleading.

GondiTilak replied:
This is now updated to use the `MiniAppJavaScriptError`

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+                self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: error.localizedDescription)
+                return
+            }
+            self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: getMiniAppErrorMessage(MiniAppErrorType.unknownError))
```
```suggestion
            self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(MiniAppErrorType.unknownError))
```

GondiTilak replied:
Sure will be using this `prepareMAJavascriptError(MiniAppErrorType.unknownError)`

rleojoseph replied:
I see the comment is not updated yet

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+            self.executeJavaScriptCallback(responseStatus: .onSuccess, messageId: callbackId, response: "SUCCESS")
+        case .failure(let error):
+            if !error.localizedDescription.isEmpty {
+                self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: error.localizedDescription)
```
```suggestion
                self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(error))
```

GondiTilak replied:
This is not working as the object `error` is of type `Error` and does not confirm to the `MiniAppsErrorProtocol`. 
For this I have now created new error enum object `UniversalBridgeError` to cater for universal bridge related errors. With this I can now use the `prepareMAJavascriptError(error)`

rleojoseph commented on Sources/Classes/core/Protocols/MiniAppBridgeDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+ Public Protocol that will be used by the Mini App to communicate
+ with the Native implementation for sending string data through interface 1 mechanism
+ */
+public protocol MiniAppBridgeDelegate: AnyObject {
```
`MiniAppBridgeDelegate` is too generic. Please change to UniversalBridgeDelegate or you can move inside `MiniAppMessageDelegate`

GondiTilak replied:
I have removed the implementation of `SendJsonToMiniApp()` from the `MiniApp.swift`.

rleojoseph commented on Sources/Classes/core/Protocols/MiniAppBridgeModel.swift:
```diff
@@ -0,0 +1,5 @@
+/// Struct is used to store the String content recieved from the Miniapps
+public struct MiniAppBridgeContent {
```
Same comment here as well

GondiTilak replied:
I have now updated the name to `UniversalBridgeContent`

rleojoseph commented on Sources/Classes/core/RealMiniApp.swift:
```diff
@@ -471,6 +471,11 @@ extension RealMiniApp: MiniAppMessageDelegate {
     }
+
+    func sendJsonToHostApp(info: MiniAppBridgeContent, completionHandler: @escaping (Result<MASDKProtocolResponse, Error>) -> Void) {
+        let error: NSError = NSError.init(domain: "MiniAppMessageBridge has not been implemented sendJsonToHostApp(info:completionHandler:) by the host app", code: 0, userInfo: nil)
+        completionHandler(.failure(error as Error))
```
```suggestion
        completionHandler(.failure(.failedToConformToProtocol))
```

GondiTilak replied:
Done the changes as suggested here.

rleojoseph reviewed (changes requested):
Thank you for the changes. Please have a look at my comments.

rleojoseph commented on CHANGELOG.md:
````diff
@@ -1,5 +1,39 @@
+<summary>Show code</summary>
+
+```swift
+func sendJsonToHostApp(info: MiniAppBridgeContent, completionHandler: @escaping (Result<MASDKProtocolResponse, Error>) -> Void) {
````
'Error' ?

GondiTilak replied:
This is updated in the later commits. Please review.

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppJavaScriptEnumerations.swift:
```diff
@@ -129,6 +130,7 @@ public enum MiniAppEvent: String {
     case resume = "miniappresume"
     case secureStorageReady = "miniappsecurestorageready"
     case secureStorageError = "miniappsecurestorageloaderror"
+    case miniappRecieveJsonString = "miniappreceivejsoninfo"
```
```suggestion
    case miniappReceiveJsonString = "miniappreceivejsoninfo"
```

GondiTilak replied:
This is now updated in the latest commit please review.

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+extension MiniAppScriptMessageHandler {
+    func sendJsonToHostApp(requestParam: RequestParameters?, callbackId: String) {
+        guard let requestParamValue = requestParam?.jsonInfo else {
+            executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: MiniAppJavaScriptError.invalidPermissionType.rawValue)
```
```suggestion
            executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(MiniAppJavaScriptError.unexpectedMessageFormat)
```

GondiTilak replied:
Changes are updated in the latest commits.

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+                self.manageJsonStringToHostAppResult(result, with: callbackId)
+            })
+        } else {
+            self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: getMiniAppErrorMessage(MiniAppJavaScriptError.valueIsEmpty))
```
```suggestion
            self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(MiniAppJavaScriptError.valueIsEmpty))
```

GondiTilak replied:
This is updated in the latest commit.

rleojoseph replied:
I'm sorry. I don't see this change yet.

GondiTilak replied:
My Bad it was missed in the previous commit will be changed in latest commit.

rleojoseph commented on Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:
```diff
@@ -745,6 +747,37 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
+    func manageJsonStringToHostAppResult(_ result: Result<MASDKProtocolResponse, Error>, with callbackId: String) {
+        switch result {
+        case .success:
+            self.executeJavaScriptCallback(responseStatus: .onSuccess, messageId: callbackId, response: "SUCCESS")
```
```suggestion
            self.executeJavaScriptCallback(responseStatus: .onSuccess, messageId: callbackId, response: MASDKProtocolResponse.success.rawValue)
```

GondiTilak replied:
This is updated with the latest commit.

GondiTilak commented on Example/Assets/en.lproj/Localizable.strings:
```diff
@@ -79,3 +79,5 @@
 "demo.app.rat.page.name.terms"                          =   "Terms";
 "demo.app.rat.page.name.list.i"                         =   "List I";
 "demo.app.rat.page.name.list.ii"                        =   "List II";
+"demo.app.rat.page.name.securestorage"                  =   "Scure Storage";
```
Updated this in the new commit.

rleojoseph commented on Sources/Classes/core/View/MiniAppViewHandler.swift:
```diff
@@ -867,3 +871,17 @@ extension MiniAppViewHandler: WKUIDelegate {
+// MARK: - Universal Bridge
+extension MiniAppViewHandler {
+    internal func sendJsonToMiniApp(string jsonString: String?) {
+        NotificationCenter.default.sendCustomEvent(
+            MiniAppEvent.Event(
+                miniAppId: appId,
+                miniAppVersion: version ?? "",
+                type: .miniappReceiveJsonString,
+                comment: jsonString ?? ""
+            )
```
You can use self.didReceiveEvent(event.type, message: event.comment) right ?

GondiTilak replied:
This will now be updated to `self.didReceiveEvent(MiniAppEvent.miniappReceiveJsonString, message: jsonString ?? "")` in the next commit.

rleojoseph reviewed (commented):
Thank you for the changes. Please have a look at my comments. Also, please make sure that the Code coverage is not dropping by 80%

GondiTilak commented on Sources/Classes/core/Protocols/UniversalBridgeDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+ */
+public protocol UniversalBridgeDelegate: AnyObject {
+    /// Interface that is used to send the string content from the Mini app. This must be implemented in the host app to receive the message from MiniApp.
+    func sendJsonToHostApp(info: UniversalBridgeContent, completionHandler: @escaping (Result<MASDKProtocolResponse, UniversalBridgeError>) -> Void)
```
Sure, got the point, I will be updating this to receive `String` object instead of the custom class object.

GondiTilak commented on Sources/Classes/core/RealMiniApp.swift:
```diff
@@ -499,3 +503,16 @@ extension RealMiniApp {
+
+extension RealMiniApp {
+    func sendJsonToMiniApp(miniAppId: String, miniAppVersion: String, jsonString: String) {
+        NotificationCenter.default.sendCustomEvent(
+            MiniAppEvent.Event(
+                miniAppId: miniAppId,
+                miniAppVersion: miniAppVersion,
+                type: .miniappReceiveJsonString,
+                comment: jsonString
+            )
+        )
```
I had missed to remove it in previous checkins, This will be removed in next checkin.

rleojoseph commented on Example/Controllers/SwiftUI/Helper/MiniAppViewMessageDelegator.swift:
```diff
@@ -170,6 +172,14 @@ class MiniAppViewMessageDelegator: NSObject, MiniAppMessageDelegate {
+
+    // MARK: - Universal Bridge support
+    func sendJsonToHostApp(info: String, completionHandler: @escaping (Result<MASDKProtocolResponse, UniversalBridgeError>) -> Void) {
+        print("Message form MiniApp: \(info)")
```
```suggestion
        MiniAppLogger.d("Message form MiniApp: \(info)")
```

laubengaier commented on Example/Controllers/SwiftUI/Features/Settings/Features/UniversalBridgeView.swift:
```diff
@@ -0,0 +1,41 @@
+struct UniversalBridgeView: View {
+    @ObservedObject var viewModel: MiniAppSettingsViewModel
+
+    @State var textToSend: String = "{\\\"data\\\":\\\"This is a sample json information from host app.\\\"}"
```
I think the string can be escaped with one backslash only 
`"{\"data\":\"This is a sample json information from host app.\"}"`

GondiTilak replied:
@laubengaier , This is not a json object it is a complete string that will be added to the json payload to be set so those backslash are needed.

laubengaier commented on Example/Controllers/SwiftUI/Features/Settings/Features/MiniAppSettingsQAView.swift:
```diff
@@ -2,82 +2,22 @@ import SwiftUI
-                }
-                .buttonStyle(.plain)
-                .padding(.vertical, 15)
+            NavigationLink {
+                MiniAppSecureStorageView(viewModel: viewModel)
+            } label: {
+                Label("Secure Storage", systemImage: "questionmark.circle")
+            }
+            NavigationLink {
+                UniversalBridgeView(viewModel: viewModel)
+            } label: {
+                Label("Universal Bridge", systemImage: "link")
```
👍

laubengaier commented on Example/Controllers/SwiftUI/Features/Settings/Features/UniversalBridgeView.swift:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    func sendDataStringToMiniApp(textString str: String) {
+        for (_, handler) in viewModel.store.handlersListDict {
+            handler.sendJsonToMiniApp?(str)
+        }
```
Just want to mention here as we discussed this works great for now but I think for a later state it would better fit the declarative approach to work with notifications or a different caching mechanism since there is a chance that the `MiniAppViewHandler` could be deallocated at random times

GondiTilak replied:
I have not seen any issues with this yet. Seem to be working as intended.

laubengaier commented on Example/Controllers/SwiftUI/Helper/MiniAppViewMessageDelegator.swift:
```diff
@@ -15,6 +15,8 @@ class MiniAppViewMessageDelegator: NSObject, MiniAppMessageDelegate {
 
     var onSendMessage: (() -> Void)?
 
+    var onSendJsonToHostApp: ((_ jsonStr: String) -> Void)?
```
this can be just `((String) -> Void)?` since the param name is omitted

laubengaier commented on Tests/Unit/MiniAppScriptMessageHandlerTests.swift:
```diff
@@ -1317,6 +1376,45 @@ class MiniAppScriptMessageHandlerTests: QuickSpec {
+                )
+                it("will return SUCCESS after host app recieves valid content strig") {
+                    mockMessageInterface.messageContentAllowed = true
+                    let command = "{\"action\":\"sendJsonToHostapp\",\"param\":{\"jsonInfo\":{\"content\":\"{\\\"data\\\":\\\"Thisisasamplejsoninformation\\\"}\"}},\"id\":\"10.050192665128856\"}"
```
same here and the other tests `\\\` can be one backslash

GondiTilak replied:
As per my comments earlier these `\\\` are needed. See below the required final json payload sent to MiniApp via JS-Bridge. `{"action":"sendJsonToHostapp","param":{"jsonInfo":{"content":"{\"data\":\"This is a sample json information\"}"}},"id":"11.172421055204431"}` . With the above `\\\` this will be the end results if we remove them then will result in parsing error.

laubengaier commented on Tests/Unit/RealMiniAppTests.swift:
```diff
@@ -743,6 +743,20 @@ class RealMiniAppTests: QuickSpec {
+                    realMiniApp.sendJsonToHostApp(info: "Test String to host from miniapp") { (result) in
+                        switch result {
+                        case .success:
+                            break
```
optional: instead of `break` you could also use `fail("should not succeed")` so the failure message is clear

GondiTilak replied:
yes I have updated this and will be pushed in next commit.

laubengaier reviewed (approved):
good job @GondiTilak! 🚀 
left a couple comments but in general lgtm so I'll approve

GondiTilak commented:
> Thank you for the changes. Please have a look at my comments. Also, please make sure that the Code coverage is not dropping by 80%

Sure, I have covered almost all the new code with the testcases and also updated and added more test cases to increase the coverage.

rleojoseph reviewed (approved):
Thank you for the changes 👍🏻 . Good job
